config.proto

```diff
@@ -127,3 +130,17 @@ message JavaAgent {
     // filter_jar_paths is the list of path to filter jars, separated by `,`
     repeated google.protobuf.StringValue filter_jar_paths = 1;
 }
+
+// CustomDataCaptureEndpoint represents a custom data capture rule for an endpoint that matches
```

Comment:
To be clear this overrides whatever you set in the DataCapture config? How do you resolve conflicting rules (e.g. 2 matching rules with different DataCapture)? All this must be specified here.

Reply:
I'm thinking we shouldn't introduce a new top level message here at all.  Instead we should move this into the existing DataCapture message.  I propose something along the lines of:

```
message DataCapture {
    Message http_headers = 1;
    Message http_body = 2;
    Message rpc_metadata = 3;
    Message rpc_body = 4;
    google.protobuf.Int32Value body_max_size_bytes = 5;
    google.protobuf.StringValue path_match = 6;
   repeated google.protobuf.StringValue headers_match = 7; 
}
```

and then change 
```
message AgentConfig {
...
    repeated DataCapture data_capture = 3;
...
}
```

I know changing `data_capture` to a repeated field is a breaking change, but it's not a widely deployed config yet, and will solve a lot of complexity in terms of defaults, precedence etc.

Reply:
If we do this then we will not have a default at all and only have rules for the matches that are given in config. What about when there is no rule given in config for a path?

Reply:
The default config would be a rule which matches everything. For the case of no config, the default in the agent would be capture everything.

Reply:
So what you are proposing @davexroth is that if user does not pass a value for DataCapture (meaning empty list) we match everything with certain values for {http|grpc}_{headers|body}. I am curious of what would happen (it's possible I think) if the user wants to disable data capture for all, I am inclined to say I would just not pass anything and not expect a default to override that. I don't feel like a repeated is intuitive.

I would go for keeping the simple config for data capture and then a repeated message, same as DataCapture but also adding a path_regex_matcher and header_regex_matcher.

Reply:
@jcchavezs I think disabling all still works pretty well with repeated. If `path_match` and `headers_match` are unspecified it means it will match all spans so disabling all would be specifying all headers and body message fields to `false` and not specifying `path_match` and `headers_match`.

Reply:
So to be clear, what we are saying is that if the user does not pass any data capture we apply the default (capture all on everything) and if the user wants to disable data capture for everything then they have to create a data capture rule with no regex and everything in off which in practice is a empty object (at least in go, zero values are false), I think that is more confusing. I honestly would prefer to be explicit on everything so having one DefaultDataCapture and then SpecificDataCapture there passing a regex for either path or header is mandatory.

Reply:
Here is an example in javaagent https://github.com/hypertrace/javaagent/pull/301

Reply:
I'm still not sure about requiring two messages, one for DefaultDataCapture and one for SpecificDataCapture - as it's the same thing.

If there's a feeling that being explicit is required, we could have two *fields*, `default_data_capture` and `data_capture` even though this can still be achieved in one field, and I'm not sure if it really add that's much clarity, and adds more combinations to worry about - what happens if `data_capturee` is defined but `default_data_capture` isn't.  Having two messages has the same problem.

Even with two fields or messages, the agent still need to define what the default behavior is if the default field/message isn't defined.

Reply:
`default_data_capture` does not have to be defined by the user because the agents generate it, The spec defines what the default behavior is.